dlt/extract/incremental.py

```diff
@@ -0,0 +1,133 @@
+from hashlib import sha1
+from typing import Generic, TypeVar, Any, Optional, Callable, Protocol, List, TypedDict, cast, Union, Tuple, get_origin, Dict
+import inspect
+
+from functools import wraps
+
+from dlt.common.json import json
+from dlt.common.typing import TDataItem, AnyFun, TFun
+from dlt.common.configuration import configspec, known_sections, with_config, resolve_configuration
+from dlt.common.configuration.specs import BaseConfiguration
+from dlt.common.configuration.specs.exceptions import NativeValueError
+from dlt.common.configuration.accessors import config
+
+
+TCursorValue = TypeVar("TCursorValue", bound=Any)
+TUniqueValue = TypeVar("TUniqueValue", bound=Any)
+LastValueFunc = Callable[..., Any]
+
+
+class IncrementalColumnState(TypedDict):
+    last_value: Optional[Any]
+    unique_keys: List[Any]
+
+
+@configspec
+class IncrementalConfigSpec(BaseConfiguration):    
+    cursor_column: str = None
+    initial_value: Optional[Any] = None
+    last_value_func: Optional[str] = None
+
+    __section__ = 'incremental'
+
+    def parse_native_representation(self, native_value):
+        if isinstance(native_value, Incremental):
+            self.cursor_column = native_value.cursor_column
+            self.initial_value = native_value.initial_value
+            self.last_value_func = native_value.last_value_func.__name__
+        else:
+            raise NativeValueError(self.__class__, native_value, 'Invalid')
+
+class Incremental(Generic[TCursorValue, TUniqueValue]):
+    
+    def __init__(
+            self,
+            cursor_column: str,
+            resource_name: str=None,
+            initial_value: Optional[TCursorValue]=None,
+            last_value_func: Optional[LastValueFunc]=None,
+            primary_key: Optional[str] = None
+    ) -> None:
+        self.resource_name = resource_name
+        self.cursor_column = cursor_column
+        self.initial_value = initial_value
+        self.last_value_func = last_value_func or max
+        self.primary_key = primary_key
+        from dlt.pipeline.current import state  # TODO: Resolve circular import
```

Comment:
we'll need to split `state,py` and move several definitions out. doable. we can go back to it when we have a working version